atlasclient/client.py

```diff
@@ -0,0 +1,268 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Python client for the MongoDB Atlas API."""
+
+import requests
+
+from atlasclient.configuration import (
+    ClientConfiguration, CONFIG_DEFAULTS as DEFAULTS)
+from atlasclient.exceptions import (
+    AtlasAuthenticationError, AtlasClientError, AtlasApiError,
+    AtlasRateLimitError)
+from atlasclient.utils import enable_http_logging, JSONObject
+
+
+_EMPTY_PATH_ERR_MSG_TEMPLATE = ('Calling {} on an empty API path is not '
+                                'supported.')
+
+
+class _ApiComponent:
+    """Private class for dynamically constructing resource paths."""
+    def __init__(self, client, path=None):
+        self._client = client
+        self._path = path
+
+    def __repr__(self):
+        return '<ApiComponent: %s>' % self._path
+
+    def __getitem__(self, path):
+        if self._path is not None:
+            path = '%s/%s' % (self._path, path)
+        return _ApiComponent(self._client, path)
+
+    def __getattr__(self, path):
+        return self[path]
+
+    def get(self, **params):
+        if self._path is None:
+            raise TypeError(_EMPTY_PATH_ERR_MSG_TEMPLATE.format('get()'))
+        return self._client.request('GET', self._path, **params)
+
+    def patch(self, **params):
+        if self._path is None:
+            raise TypeError(_EMPTY_PATH_ERR_MSG_TEMPLATE.format('patch()'))
+        return self._client.request('PATCH', self._path, **params)
+
+    def post(self, **params):
+        if self._path is None:
+            raise TypeError(_EMPTY_PATH_ERR_MSG_TEMPLATE.format('post()'))
+        return self._client.request('POST', self._path, **params)
+
+    def delete(self, **params):
+        if self._path is None:
+            raise TypeError(_EMPTY_PATH_ERR_MSG_TEMPLATE.format('delete()'))
+        return self._client.request('DELETE', self._path, **params)
+
+    def get_path(self):
+        return self._path
+
+
+class _ApiResponse:
+    """Private wrapper class for processing HTTP responses."""
+    def __init__(self, response, request_method, json_data):
+        self.resource_url = response.url
+        self.headers = response.headers
+        self.status = response.status_code
+        self.request_method = request_method
+        self.data = json_data
+
+    def __repr__(self):
+        return '<{}: {} {}, [HTTP status code: {}]>'.format(
+            self.__class__.__name__, self.request_method,
+            self.resource_url, self.status)
+
+
+class AtlasClient:
+    """An easy-to-use MongoDB Atlas API client for Python. """
+    def __init__(self, *, username, password, base_url=None,
+                 api_version=None, timeout=None, verbose=None):
```

Comment:
Suggest adding defaults for these kwargs, eg `base_url=DEFAULTS.BASE_URL`, etc...